akka-stream/src/main/scala/akka/stream/javadsl/DelayStrategy.scala

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (C) 2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.javadsl
+
+import akka.stream.scaladsl
+import akka.util.JavaDurationConverters.JavaDurationOps
+
+import scala.concurrent.duration.FiniteDuration
+
+/**
+ * Allows to manage delay and can be stateful to compute delay for any sequence of elements,
+ * all elements go through nextDelay() updating state and returning delay for each element
+ */
+trait DelayStrategy[T] {
+
+  /**
+   * Returns delay for ongoing element, `Duration.Zero` means passing without delay
+   */
+  def nextDelay(elem: T): java.time.Duration
+
+}
```

Comment:
AFAIR a companion object to a trait will not be accessible from Java, please add a Java test to verify that the factory methods are actually callable. Should be fixable by making it an abstract class instead. (The test can just verify that it compiles, no need to duplicate the tests for the actual logic IMO)

Reply:
I see. btw, for my own curiosity, any reason why javadsl is not written in Java?

Reply:
I think we settled on it out of convenience mostly, even just writing the APIs in Java is quite a bit more verbose and make it harder to interact with Scala code from the impl unless you do only public API in Java and impl in Scala. Using Scala to write the entire Java APIs otoh requires some extra thought, but that's what we do across Akka.

Reply:
(we dropped support for scala 2.11)